### ISSUE TITLE
Sensible calculation of social insurance contributions

### ISSUE THREAD

**Issue by hmgaudecker** [iza-institute-of-labor-economics/gettsim#332]
### Current and desired situation

Currently, we are defining variables `regulär_beschäftigt` and others, calculate overall social insurance contributions, then adjust if mini or midi Job.

Adds a lot of unnecessary code and will be confusing once we are better with modelling civil servants, self-employed, etc.

### Proposed implementation

Calculate directly as one function per social insurance type using the piecewise polynomial machinery.

**Comment by ChristianZimpelmann** [comment]
As the calculation of contributions for mid job is fairly complicated I am afraid we cannot use the piecewise polynomial machinery here.

@hmgaudecker, can you provide more detail? Otherwise, I would close this issue.

**Comment by hmgaudecker** [comment]
Well, that was one idea only, thinking that the basic form of social insurance contributions is that of a deductible (mini-jobs), a phase-in period (midi), and a fixed rate up to a threshold. What is wrong with that thinking?

In any case, we should determine based on "einkommen_abhängige_beschäftigung" or so (bruttolohn currently, may need to distinguish more for civil servants), rather than this "regulär_beschäftigt" thing.

**Comment by ChristianZimpelmann** [comment]
> Well, that was one idea only, thinking that the basic form of social insurance contributions is that of a deductible (mini-jobs), a phase-in period (midi), and a fixed rate up to a threshold. What is wrong with that thinking?

That's true in general, but the phase-in period (midi) is quite complicated (the sum of AG+AN contribution is calculated on a reduced Bemessungsentgelt, but the AG contribution is based on the full income. The AN contribution is then the difference). I don't think the piecewise polynomial machinery as it is currently implemented can be used for that.

> In any case, we should determine based on "einkommen_abhängige_beschäftigung" or so (bruttolohn currently, may need to distinguish more for civil servants), rather than this "regulär_beschäftigt" thing.

But `regulär_beschäftigt` is currently calculated based on `bruttolohn_m`. I think it makes sense to not make the comparison between `bruttolohn_m` and `soz_vers_beitr_params["geringfügige_eink_grenzen_m"]["midi_job"]` in several functions directly, but use `regulär_beschäftigt`.

**Comment by hmgaudecker** [comment]
Again, I do not care much whether in the end we will wind up using the piecewise polynomial machinery. The solution might well be different than my initial thought, but the problem remains, so there is no reason to close the issue.

- AFAICT, we repeat the same code for all types of social insurance contributions. That should definitely change.
- The construct `regulär_beschäftigt` does not work if someone has a mini-job on the side, the naming does not work if we add civil servants 
- What you write above might suggest calling `piecewise_polynomial` multiple times (AG+AN, AG only), but we'll have to work through the exact structure first. Would also need to make sure that we cover the special cases of health insurance and long term care insurance.

**Comment by hmgaudecker** [comment]
Done.
